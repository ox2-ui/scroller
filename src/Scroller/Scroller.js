import PropTypes from 'prop-types';
import React, { Component } from 'react';

const styles = {
  scroller: {
    position: 'absolute',
    top: '0',
    left: '0',
    right: '0',
    bottom: '0',
    overflowY: 'auto',
    overflowX: 'hidden',
    WebkitOverflowScrolling: 'touch',
    paddingBottom: '40px',
    backgroundColor: 'transparent',
  },
};

const FIRST_ELEMENT = 0;

/**
 * Scroller Component
 */
class Scroller extends Component {
  static propTypes = {
    /**
    * Elements to be rendered inside scroller
    */
    children: PropTypes.node.isRequired,
    /**
    * The css class name of the root element.
    */
    className: PropTypes.string,
    /**
    * Override the inline-styles of the root element.
    */
    style: PropTypes.object,
  };

  static defaultProps = {
    className: '',
    style: {},
  };

  state = {
    scrollStartY: 0,
  };

  handleTouchStart = event => {
    this.setState({
      scrollStartY: event.touches[FIRST_ELEMENT].clientY,
    });
  };

  handleTouchMove = event => {
    const scrollMoveY = event.changedTouches[FIRST_ELEMENT].clientY;
    const topOffset = event.currentTarget.scrollTop;
    const maxContentHeight =
      event.currentTarget.scrollHeight;
    const itemHeight = event.currentTarget.clientHeight;
    // Prevent dragging when cotent bottom is reached
    if (this.state.scrollStartY > scrollMoveY) {
      if (topOffset + itemHeight === maxContentHeight) {
        event.preventDefault();
      }
    } else {
      // Prevent dragging when 'top' offset is 0 (content top reached)
      if (!topOffset) {
        event.preventDefault();
      }
    }

    this.setState({
      scrollStartY: event.touches[FIRST_ELEMENT].clientY,
    });
  };

  render() {
    const {
      className,
      children,
      style,
    } = this.props;

    const mergedStyles = {
      ...styles.scroller,
      ...style,
    };

    return (
      <div
        className={className}
        onTouchMove={this.handleTouchMove}
        onTouchStart={this.handleTouchStart}
        style={mergedStyles}
      >
        {children}
      </div>
    );
  }
}

export default Scroller;
