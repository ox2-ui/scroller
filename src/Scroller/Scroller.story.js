import React from 'react';
import { storiesOf } from '@storybook/react';
import Scroller from './Scroller';

const style = {
  container: {
    height: '100px',
    position: 'relative',
    backgroundColor: '#e3e3e3',
  },
};

storiesOf('Scroller', module)
  .add('short', () =>
    <div style={style.container}>
      <Scroller className="color:backdrop padding:7 ">
        <div>
          Lorem ipsum dolor sit amet, consectetur
          adipisicing elit. Nihil repellendus officiis saepe
          debitis, sapiente, quas vel veniam nesciunt. Autem
          ipsa incidunt minus, eligendi consectetur id est
          vel magni harum doloremque. Voluptatibus,
          perferendis, corrupti! Autem enim molestiae
          excepturi sequi fugiat, iste sit ab. Ipsum in
          quae, nihil maxime. Facilis ducimus dolorem
          exercitationem iusto ex reprehenderit praesentium
          rem, enim quo eum accusamus molestiae laudantium
          minus recusandae nam, deleniti quam voluptatibus.
          Repellat dolorum facilis, consequuntur sint
          inventore, asperiores minima architecto earum
          laborum nisi tenetur totam error, alias blanditiis
          qui quisquam aliquid aperiam consequatur suscipit
          harum velit sed at! Repudiandae aspernatur,
          exercitationem incidunt atque.
        </div>
      </Scroller>
    </div>,
  )
  .add('long', () =>
    <div style={style.container}>
      <Scroller className="color:backdrop padding:7">
        <div>
          {' '}Lorem ipsum dolor sit amet, consectetur
          adipisicing elit. Voluptatem nobis quia corporis
          iste rerum odio laboriosam, animi ad saepe
          praesentium nisi, eius itaque impedit, quos error
          modi doloribus, quam at accusamus repellendus
          officiis sint illo molestias deleniti! Libero,
          asperiores quos esse quia, ab eligendi beatae
          quaerat porro dignissimos aliquam obcaecati quae
          cupiditate accusantium explicabo soluta velit
          minus. Qui earum distinctio corporis? Blanditiis
          placeat ab vero esse perferendis cumque veniam
          architecto nobis, dolor quis! Alias deleniti
          adipisci, cupiditate illum doloribus quo aut
          libero voluptatem quisquam vitae facere veritatis
          illo velit pariatur error iste quasi ad dolores
          repellat itaque iure. Eius fuga repudiandae
          maiores voluptas aliquid, expedita, voluptates
          laudantium quidem officia inventore quasi magni
          delectus aut sed cum! Ullam minima, odit natus
          repellendus architecto omnis, non pariatur quo
          aliquam in, quas beatae. Laudantium assumenda, ea
          debitis. Quisquam ab natus porro soluta, repellat
          autem laboriosam necessitatibus fugiat, aperiam
          esse quos adipisci officiis alias vitae odit
          dignissimos. Rem et voluptatum labore,
          consequuntur minima officiis ut adipisci commodi
          fuga, quibusdam deleniti repellendus magnam. Esse
          explicabo excepturi voluptates qui ab laborum
          provident fuga ipsum dicta deserunt libero,
          eveniet eius officia praesentium perspiciatis
          veritatis soluta natus vitae corrupti, aperiam
          dolores illo? Iusto asperiores quaerat impedit
          ipsam id!
        </div>
      </Scroller>
    </div>,
  );
