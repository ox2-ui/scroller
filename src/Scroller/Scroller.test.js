/* eslint-disable no-unused-vars */
import React from 'react';
import { shallow, render, mount } from 'enzyme';
import Scroller from './Scroller';

const handler = () => true;

it('renders correctly', () => {
  const wrapper = mount(
    <Scroller>
      <div>
        Lorem ipsum dolor sit amet, consectetur
        adipisicing elit. Nihil repellendus officiis saepe
        debitis, sapiente, quas vel veniam nesciunt. Autem
        ipsa incidunt minus, eligendi consectetur id est
        vel magni harum doloremque. Voluptatibus,
        perferendis, corrupti! Autem enim molestiae
        excepturi sequi fugiat, iste sit ab. Ipsum in
        quae, nihil maxime. Facilis ducimus dolorem
        exercitationem iusto ex reprehenderit praesentium
        rem, enim quo eum accusamus molestiae laudantium
        minus recusandae nam, deleniti quam voluptatibus.
        Repellat dolorum facilis, consequuntur sint
        inventore, asperiores minima architecto earum
        laborum nisi tenetur totam error, alias blanditiis
        qui quisquam aliquid aperiam consequatur suscipit
        harum velit sed at! Repudiandae aspernatur,
        exercitationem incidunt atque.
      </div>
    </Scroller>,
  );

  expect(wrapper).toMatchSnapshot();
});
