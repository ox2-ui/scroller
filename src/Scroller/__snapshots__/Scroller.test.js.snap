// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Scroller
  className=""
  style={Object {}}
>
  <div
    className=""
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    style={
      Object {
        "WebkitOverflowScrolling": "touch",
        "backgroundColor": "transparent",
        "bottom": "0",
        "left": "0",
        "overflowX": "hidden",
        "overflowY": "auto",
        "paddingBottom": "40px",
        "position": "absolute",
        "right": "0",
        "top": "0",
      }
    }
  >
    <div>
      Lorem ipsum dolor sit amet, consectetur adipisicing elit. Nihil repellendus officiis saepe debitis, sapiente, quas vel veniam nesciunt. Autem ipsa incidunt minus, eligendi consectetur id est vel magni harum doloremque. Voluptatibus, perferendis, corrupti! Autem enim molestiae excepturi sequi fugiat, iste sit ab. Ipsum in quae, nihil maxime. Facilis ducimus dolorem exercitationem iusto ex reprehenderit praesentium rem, enim quo eum accusamus molestiae laudantium minus recusandae nam, deleniti quam voluptatibus. Repellat dolorum facilis, consequuntur sint inventore, asperiores minima architecto earum laborum nisi tenetur totam error, alias blanditiis qui quisquam aliquid aperiam consequatur suscipit harum velit sed at! Repudiandae aspernatur, exercitationem incidunt atque.
    </div>
  </div>
</Scroller>
`;
